compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BTupleType.java

```diff
@@ -138,20 +139,24 @@ public void setIntersectionType(BIntersectionType intersectionType) {
 
     // In the case of a cyclic tuple, this aids in
     //adding resolved members to a previously defined empty tuple shell in main scope
-    public boolean addMembers(BType memberType) {
+    public boolean addMembers(BTupleMember memberType) {
         // Prevent cyclic types of same type ex: type Foo [int, Foo];
-        if (memberType instanceof BTupleType && ((BTupleType) memberType).isCyclic &&
-                memberType.getQualifiedTypeName().equals(this.getQualifiedTypeName())) {
+        if (memberType.type instanceof BTupleType && ((BTupleType) memberType.type).isCyclic &&
+                memberType.type.getQualifiedTypeName().equals(this.getQualifiedTypeName())) {
             return false;
         }
         this.tupleTypes.add(memberType);
-        if (Symbols.isFlagOn(this.flags, Flags.READONLY) && !Symbols.isFlagOn(memberType.flags, Flags.READONLY)) {
+        if (Symbols.isFlagOn(this.flags, Flags.READONLY) && !Symbols.isFlagOn(memberType.type.flags, Flags.READONLY)) {
             this.flags ^= Flags.READONLY;
         }
-        setCyclicFlag(memberType);
+        setCyclicFlag(memberType.type);
         return true;
     }
 
+    public boolean addMembers(BType memberType) {
+        return this.addMembers(new BTupleMember(memberType));
+    }
```

Comment:
Given that this is called (only) from the `ImmutableTypeCloner`, do annotations still work as expected when you have a `readonly` intersection with a tuple with annotations?

Reply:
Should Intersection types of readonly tuples be special-cased in `AnnotationDesugar` or should we change how immutable tuple types behave ?

Reply:
I believe we need to revisit annotations with intersection types. 

For now though, can't we create a new tuple member in the immutable type cloner anyway? And use the preceding method instead of introducing this?